### ISSUE TITLE
Ability to describe files that belong together

### ISSUE THREAD

**Issue by hannes-ucsc** [HumanCellAtlas/metadata-schema#623]
Supersedes #579 

### Use case: Direct matrix download

A user of the data browser downloads an expression matrix from an analysis bundle. The matrix is stored in those bundles as a zarray store, a collection of tens or hundreds of smaller files. Each of these files is a separate file entity in the data store. Without special intervention, each of these files would show up as a separate row in the Files tab of the data browser and the user would have to tediously download them individually if they wanted to directly download an complete expression matrix. The proliferation of files is currently addressed via a hard-coded rule in Azul (the data browser indexer) to suppress all but one file from a zarray store: the `.zattrs` file which should be unique per store. With this rule in place, the user is not able to download a complete zarray store, even if they were determined enough to download each file individually. I initially proposed replacing the hard-coded rule with a metadata-driven approach (#579) but that proposal doesn't address the download use case either.

### Use case: When downloading a `.bam`, also download the corresponding `.bai`

When accessing a `.bam`, users typically want to access the correponding `.bai` index file because the index is small and takes longer to create than to download. When accessing BAM data selectively (aka "slicing"), the index is a prerequisite. The convention is for the `.bai` and `.bam` file to share the same base name. 

To find the `.bai` file for any given `.bam` file in a bundle, one would have to apply this convention by scanning file names listed in the bundle's manifest.  To do the same interactively using the data browser one would have to filter for `bai` and `bam` and sort the Files tab by file name,  but this is not reliable because two different bundles could have distinct bam/bai pairs that use the same base name. 

We might want to facilitate these use cases by using metadata to express the relationship between a `.bai` and the `.bam` it was generated from.

### Use case: Express relationships between imaging files

**IMAGING INPUT**
Imaging does not require this full list of imaging right now because the json files are self referencing. All the files below under experiment.json need to exist as one group. The one experiment.json is the top level file that should be displayed.

**IMAGING OUTPUT**
Will contain zarrays generated by analysis so this use case is identical to the sequencing case.

**FYI**
#### SpaceTx
Imaging files from spacetx have a hierarchy TBD but similar to:
```
experiment.json
    codebook.json
    version.json
    data_manifest
        version.json
        field_of_view.json
            field_of_view_coordinates (this may be a zarray)
            field_of_view_indices (this may be a zarray)
    nuclei_manifest
        version.json
        codebook.json
        field_of_view.json
            field_of_view_coordinates (this may be a zarray)
            field_of_view_indices (this may be a zarray)
```
An 'is referenced by' link (or similar) between files would allow the browser to understand this hierarchy and choose how to display files depending on situational requirements.

**Comment by hewgreen** [comment]
One implementation for this could be a new field in file_core.json (NB analysis_file.json imports file_core.json) containing an array of `sub file names'. I think this would allow file nesting which could be used to power the browsers requested functionality. But it does push the burden onto the submitter/wrangler (short term, unstable) and pipeline developers (longer term, more stable) to maintain these links. By putting this field on file_core.json, wranglers and submitters will have the option to edit the field (this is useful while manually wrangling imaging datasets) but longer term this may be best moved to the analysis_file.json.

**Comment by malloryfreeberg** [comment]
> this may be best moved to the analysis_file.json

Wouldn't it have to be in any imaging data files we get, as well?

**Comment by malloryfreeberg** [comment]
@hannes-ucsc @diekhans to describe their proposed implementations (to be discussed on future call).

**Comment by diekhans** [comment]
A possible approach would be to add a file group guid that is associated with all of the files that are linked together.  Given one file, all files in a group can easily be found.

This could be used in conjunction with putting all related files in a directory, providing a formal way to find the group other than name matching.

This has the disadvantage that the file group is not a first class object. Making explicit queries by name for group rather clunky.

Does anyone know why directories are forbidden?  It seems like a natural way to group files of all sorts.

**Comment by lauraclarke** [comment]
I like the idea of some sort of group id but I am concerned about situations like imaging files where non-expert submitters might need to define them in large numbers

I am afraid if we choose to use directories and no explicit piece of information in the metadata about how things are groups this information gets lost when people download the files (which will happen)

**Comment by diekhans** [comment]
Good points @lauraclarke.   A straightforward way for users to define groups is to put them in a directory and then have the ingest software define the group id.  Downloading all files to a directory would be very useful for the user.

making directories first-class objects would allow for more complex groupings in the future, which group ids would not.

**Comment by malloryfreeberg** [comment]
@lauraclarke @diekhans I believe making directories first-class objects would require changes to the upload service (which currently strips all but the file name when a file is uploaded), possibly the data store and secondary analysis pipelines. Intuitively I also like the idea of grouping related files using a directory structure (the directory name could suffice as a "group ID"), but I'm aware that this is a concept not currently thought about from an ingest perspective.

> Does anyone know why directories are forbidden? 

I'm not sure about the answer to this question, but a good place to ask might be the #tech-architecture channel. I can post a Q in the channel and see if we get any clarity.

If we do put files in directories, this still doesn't sufficiently address the use case of what files to show in the Data Browser. It gets us partly there, but we still can't answer the question of "which of the tens or hundreds of smaller files in a zarray store should be viewed in the Data Browser for a user to select and download?".

**Comment by hewgreen** [comment]
Maybe the ID doesn't need to be a uuid. Elsewhere in the metadata we have used group IDs in a similar way e.g. library prep ID, Biomaterial ID's and Protocol ID's. We could capture user supplied IDs that only need to be unique within a submission. Then with submission uuid and this string you can work out how to nest files. The advantage is that if a wrangler needs to fill these out manually they could do this quite simply, equally analysis can generate IDs. The downside is that it is not unique across multiple submissions without also using the submission uuid.

**Comment by hannes-ucsc** [comment]
> Does anyone know why directories are forbidden?

I think it was a security concern. To prevent an attacker from putting `/etc/passwd` or `../../../../../../etc/passwd` into a bundle, cheat a user into downloading the bundle, which would overwrite `/etc/passwd` with the file bundled by the attacker, and then login to their system.

But the restriction could be relaxed to only allow relative paths and only those without `..` and `.` elements.

**Comment by diekhans** [comment]
even .. can be allowed as long as it checks from not going out of the root of submission.

Yes, directories could be useful for many things and make the user a lot happier when  they download a larger

**Comment by hannes-ucsc** [comment]
To summarize, there are the following options on the table:

- (a) Introduce a `group` property associated with each `file` entity (the core type) whose value is
  - (a1) a UUID
  - (a2) an identifier unique to the submission, potentially hand-picked by a submitter
- (b) Support directories in bundles by allowing `/` in the `name` property of bundle manifest entries and
  - (b1) Describe each file in the directory with a separate metadata entity, like `analysis_file`.
    - (b1a) Mark the directory as a group with an empty `.group` file to differentiate the directory from potential other, non-grouping directories
  - (b2) Describe the directory (!) with a single metadata entity as if it were a file

Option b2 is interesting because it is economical with metadata. The SS2 analysis bundles currently in staging and production have an `analysis_file` file entity per file in the zarr store. Those `analysis_file` entities bloat the bundle, slow down the ingestion and indexing process and add no informational value for the user.

Option b2 also allows for directories that aren't groups of files. If you want to just put files in a directory, you do so by prefixing the name with `somedir/` but describe each file with its own metadata entity.

**Comment by diekhans** [comment]
For (a), my thinking is a UUID and a user-friendly name for the group would meet both management needs of having global unique ids and way to describe it in in a gui.  This does mean duplicating data.

(b2) is the proved method of implementing directories in operating systems.  It provides a single place for any time of metadata (user-friendly names) and makes hierarchical traversal easy.,  It could be optional for directories that don't have any additional grouping semantics.

So I would favor b2.   It all cases, we need to make it easy for the user.

**Comment by hewgreen** [comment]
New AI from metadata meeting. We need to write up a full spec around for option b2 and put this on the technical architecture call for discussion. @diekhans

**Comment by claymfischer** [comment]
Very excited to see some traction on this, B2 seems like a wonderful solution with the caveat @hewgreen and @lauraclarke brought up (user-uploaded hierarchy is not necessarily meaningful to others).

**Comment by diekhans** [comment]
Yes, when I suggested user upload hierarchy, I was thinking of forcing the user to meet the upload requirements, not taking any old user hierarchy.  

I am not a typical user and am happy to not do this for upload.

**Comment by hewgreen** [comment]
Seeing the cost benefit analysis of creating another entity will be very interesting. My uninformed guess at this point would be that it would be too high and adding a grouping id to metadata in a filed on file would the easier fallback. I'm looking forward to seeing how this comes out.

**Comment by joshmoore** [comment]
Trying to get caught up with this thread after a ping from @hewgreen, here are some thoughts/comments from the world of imaging. Sorry for the brevity, I'm happy to go into more depth or join a call:

> Description: to suppress all but one file from a zarray store: the `.zattrs` file which should be unique per store.

If I'm understanding, there's an issue here: in some zarr filesets, there will be multiple .zattrs file, since a single file can contain multiple arrays (contained under potentially hierarchical groups)


> Mallory: which currently strips all but the file name when a file is uploaded
> Matt: I'd advocate stripping any user provided structure and applying any nesting we decide systematically over all the datasets so that this grouping metadata is the same system wide.
> Mark: I was thinking of forcing the user to meet the upload requirements, not taking any old user hierarchy.

I think the user structure is going to be important and follows from the "potentially hierarchical groups" zarr issue. To explain: a worst case scenario in the imaging world is that the user provides a directory which contains `E>1` imaging experiments, each of which contains `I>1` multi-dimensional images. I'd imagine it's feasible to disallow the former ("user must extricate the filesets from one another") but dropping the hierarchy of a single fileset (or changing directory or file names, etc) is likely to corrupt the internal structure of the fileset.

Regarding the solutions, having a group object (what I call a fileset) should work but the contained files would need to maintain their slashes (probably standardizing on posix `/`). Having a directory object would also work, but a directory would need to have an optional parent directory with the top of that hierarchy being the equivalent of the group.

From the UI perspective, when it comes to an imaging experiment, it should be possible to tell ingest/upload reliably which files are clearly part of the experiment and therefore can be collapsed visually into a single 📷 icon. There may additionally be "attached" files which could be handled separately. (This process could be kept simple or be done recursively as in the osx app example in which an internal group could additionally have files with their own structure.)

**Comment by hewgreen** [comment]
Note: I have amended the imaging portion of this ticket after conversations with imaging folks. See above for a description of the requirements. The requirement is simpler than first envisioned.

@diekhans and I have discussed what containers exist and where immutable units lie. In conclusion, a process entity from final biomaterial to multiple files (aka the current form of a bundle) as a unit of related files is relied upon to link R1 R2 I1 I2 files in SS2 and 10X experiments. Equally the vision for an **ingested input** for imaging (a fileset) is one experiment.json per bundle and so the bundle would link all files under experiment.json (as described at the top of the issue). Note that the hierarchy of an image fileset under the experiment.json is mostly self linking so we do not need to provide additional nesting information for this.

Therefore, bundle as a container works at the moment. The only place it falls short is for zarr files. This applies to both imaging and sequencing datasets and needs to be addressed in this ticket.

Another important clarification is that zarr files do not need to be provided by any users (imaging or sequencing) at this moment in time. Therefore, we only need to support this file grouping for programatic submission from green pipelines (which I'm taking to include starfish at some point in the future).

As per Hannes/Josh's requirements, multiple subsets of files within the bundle container need to be grouped (**requirement 1**). In addition, it is sometimes appropriate for one of the files within that group to represent the group; as is the case for zarr (**requirement 2**). This representative file is what Hannes wants to show users in the data browser.

I've tried to suggest a balanced solution that is generic but also limits the scope to suit our current requirements. I propose the following:

### To meet requirement 1: add a new grouping id field to the `type/file/analysis_file.json`

This metadata field will group files together by using the same ID. This does not need to go on file core as we will only trust the sub grouping provided programatically. We do not need to take grouped files from submitters.

An alternative to this implementation mentioned by (possibly not championed by) @diekhans would be to make a new entity ~'analysis_file_group' and nest file entities under there. This may be nicer modelling but extends the model placing some burden on the submitter. Whilst this is only programatic this may not be an issue.

@hannes-ucsc may be best placed to say whether this would be more or less difficult to consume as a field or entity.

### To meet requirement 2: provide enum for types of file content to identify representative files

This is being discussed in [another ticket 542](https://github.com/HumanCellAtlas/metadata-schema/issues/542) as there are other reasons for this requirement. Note that this will address the point @joshmoore makes. This will allow multiple sub groups of zarrs per bundle.

The data browser will be able to collapse zarr using the new group ID and also collapse an imaging experiment in the same way bundles are devised.

### Missing

This solution does not allow complex multiple nesting. As an example, we won't be able to collapse FOV within an experiment without inspecting the imaging data files themselves. From my understanding we don't need this at the moment. If we did I think a '/' file path implementation may be required as mentioned by @joshmoore and others.


The following people may also want to comment:
@lauraclarke @malloryfreeberg @tburdett @ambrosejcarr @simonjupp @justincc @claymfischer

**Comment by justinccdev** [comment]
The only comment I will make at this time is that I think we want to try and avoid tying this into bundles if at all possible. What if some other type of analysis also has `experiment.json`?

**Comment by tburdett** [comment]
Catching up on this discussion after a mention. I understand the need to download groups of files together, but I don't think I have fully grasped what @joshmoore calls 'filesets' actually represents; and what defines an imaging experiment? (I probably know this but the context switch is wrong-footing me so I could use a simple summary!)

@hewgreen bundles are a meaningless concept here; they have no ability to describe relationships between files, and there are no semantics attached to why multiple files may occur in a single bundle: you need the metadata to do this for you.

The DCP metadata model currently provides one and only one mechanism for "grouping" files: related _files_ are all derived from the same _biomaterial_ via a single _assay process_ or the same _file_ via a single _analysis process_. 

This representation is sufficient to handle groupings of raw sequence files for SS2 and 10x as well as BAM/BAI outputs from analysis pipelines. In both cases a flat list of the files that make up the output of the process works fine. The metadata model is designed to allow deeper expressivity - e.g. relationships between files - to be added if needed, although we have not had this usecase before.

Is the standard mechanism for grouping (i.e. a flat list of files derived from a single process, in this case an _imaging process_, plus associated descriptive metadata) not sufficient for the imaging usecase?

**Comment by hannes-ucsc** [comment]
There is strong resistance against the elimination of bundles. I heard a 2nd hand account of a proposed solution to the bundle size limit but I wasn't impressed.

I'm not sure we NEED to solve the imaging use case here. Using unbundled DSS files for zarrays in conjunction with hashing intra-zarray file paths to v5 UUIDs for identifying files in the DSS seems like the most promising solution to my sore eyes.

@tburdett, correct me if I am wrong: the primary key for metadata entities (provenance.document_id) is only a UUID. No versioning information is persisted in the document itself. Nor is it persisted in Ingest's database. When an entity is updated, a new version of the corresponding DSS file is created. The DSS is therefore the only place where previous versions of the entity exist.

If the above is correct, it would be a good argument FOR bundles (as transactions) because they represent the only available immutable snapshot of the metadata graph (or fragments of said graph) at any given point in time.

**Comment by hannes-ucsc** [comment]
@joshmoore, good point about .zattrs not being unique.

**Comment by diekhans** [comment]
It is interesting that this ticket is about coming up with an
entity model to cleanly and accurately represent the data we
wish to store.

Yet some how it has circled back to bundles.  If our
`transaction' (bundle) mechanism is so heavy weight that we need
to hack around it to store the data outside of bundles, then we
really need to address the underlying mechanism rather than make
the system even more complex.

Speaking of versioning, I am trying to reconstruct the versioned
graph, it is not an easy task.  You can't do it from the
metadata alone.  Oddly, the download of a bundle doesn't include
the bundle definition.

**Comment by joshmoore** [comment]
> Is the standard mechanism for grouping (i.e. a flat list of files derived from a single process, in this case an _imaging process_, plus associated descriptive metadata) not sufficient for the imaging usecase?

@tburdett : no, unfortunately. The richness of the formats requires the directory hierarchy to express various relationships. A "fileset" is the set of directories and files which represent a single acquisition (the output of the imaging process). And to complicate matters, in most cases, I assume the knowledge of this fileset structure is encoded in Bio-Formats.

**Comment by tburdett** [comment]
@joshmoore that's a really clear explanation, thanks. So, to check my understanding - the metadata entity model is expressive enough to group _files_ resulting from a single _acquisition process_. Additional metadata (specific to a particular methodology e.g. SpaceTx) is required to describe the relationships between the _files_ in a _fileset_ that result from that acquisition.

Is this the metadata that is stored in `experiment.json`?

**Comment by tburdett** [comment]
> There is strong resistance against the elimination of bundles

@hannes-ucsc I perhaps wasn't too clear - I'm not advocating the elimination of bundles, just commenting that for this scenario they don't give us any mechanism to describe groupings for files. As long as we can describe how files are grouped, we may choose to write them to the DSS in bundles that reflect those grouping (or we may not), but it won't solve the underlying representational problem.

**Comment by tburdett** [comment]
> the primary key for metadata entities (provenance.document_id) is only a UUID. No versioning information is persisted in the document itself

@hannes-ucsc this is correct, yes.

**Comment by hewgreen** [comment]
The word 'bundle' is just used here to refer to a specific granularity of slice from metadata graph. I use it because its easier than always saying 

> related _files_ are all derived from the same _biomaterial_ via a single _assay process_ or the same _file_ via a single _analysis process_.

or as I put it:

> a process entity from final biomaterial to multiple files

This conversation does not need to be about bundles as stored in the DSS but that granularity of slice in the metadata that happens to be the same granularity as a bundle. The term 'assay process' has ambiguity issues with lack of scientific context but we can pick that term back up if we like it better than 'bundle'.

This granularity is relevant because, as Tony mentioned, it's the container for file grouping we currently have. We need to model a more granular subset of files for both the sequencing use case and the imaging one. I think this is just a modelling tweak.

---

I have prepared this document to better outline the problem so we can reach consensus on an implementation.

https://docs.google.com/document/d/1PLRp2I1u_nzz6aKX-nE30maeqhkNCsHKEwTpI8o1lIQ/edit#

**Comment by joshmoore** [comment]
> Additional metadata (specific to a particular methodology e.g. SpaceTx) is required to describe the relationships between the _files_ in a _fileset_ that result from that acquisition. Is this the metadata that is stored in `experiment.json`?

@tburdett : where Bio-Formats is concerned, the fileset when detected will choose itself such a key file. In the SpaceTx case, we can either detect this file via a content heuristic (what keys are present) or a filename convention ("experiment.json"), but yes, usually there's some top-level file which defines the relationship between the files. (There _are_ however examples of formats where only the sum of file _patterns_ defines the relationship, but I suggest we avoid that for a long as possible.)

**Comment by tburdett** [comment]
OK, thanks @joshmoore. So far I'm not seeing a need to change the HCA metadata entity model - _files_ are still fundamentally grouped based on being multiple outputs of the same _acquisition process_. But consumers and downstream rely on the _fileset_ directory structure so capturing this critical. Fair?

Would it be sane/useful for the metadata team to add a new metadata schema (under https://github.com/HumanCellAtlas/metadata-schema/tree/master/json_schema/type/file) to represent the concept of such a _key file_, such that downstream software would be able to look up keys from the metadata rather than than relying on e.g. a filename convention ("experiment.json")?

**Comment by joshmoore** [comment]
> But consumers and downstream rely on the _fileset_ directory structure so capturing this critical. Fair?

Yes.

> to represent the concept of such a _key file_


It might be possible to capture this in (mime)type information for the key file; an alternative though if the modeling already supports it is for the fileset to be _ordered_ and consider the first file key.

**Comment by malloryfreeberg** [comment]
Noting here that there is now a Data Store proposal to support virtual paths ("/") in filenames: https://github.com/HumanCellAtlas/data-store/issues/1885

**Comment by malloryfreeberg** [comment]
It seems like this issue has been superseded by #880? Should we close this issue?

In parallel we are discussing "describing the contents of a file" #542. Some of the issues here/#880 might be helped by #542.

**Comment by diekhans** [comment]
Note that is related to, but also may be different than #880.  #880 describes an archive stored as a directory tree where the contents are opaque to the user. #623 could cover cases where the files are grouped, but the individual files and their meaning is transparent to the users, such as a pair of FASTQ files.

**Comment by tburdett** [comment]
My read: between the metadata model that already exists (i.e. a single process generates multiple discrete files) and #880 (a single process generates a directory structure representing a single archive) we have covered the usecases described in this ticket

**Comment by justincc** [comment]
@diekhans Could you summarize what the proposal now is?

**Comment by diekhans** [comment]
Summary:
1. The metadata file_core entity is changed to allow representing complex data types, such as zarr, which are stored as multiple files in a directory hierarchy as a single entity.  This eliminates the need to store a large number of entities actually represent the internals of a data structure. 
1. Checksums are removed from the file entity.  All file integrity management, including client side, becomes the responsibility of the DSS.
1. The metadata file entities contain the FQID of the file or directory
1. The submitter and consumer see the directory for archive as a single data item, named by the directory.
1. There are many issues related to ingest and DSS that need to be investigated
1. The priority is unclear

**Comment by malloryfreeberg** [comment]
@diekhans Are you happy to be the "shepherd" for this functionality? I think there are a few things needed for this, and it's not clear where we are with each thing being proposed, decided, etc.

**Comment by diekhans** [comment]
I believe there are  two use cases here:

1.   A small number of which are where their association is known to most users, such as paired FASTQs, or BAM and BAM index files.
1.  A directory that is a container for files, such as zarr

**Comment by daniwelter** [comment]
Additional use case raised by @samanehsan: #983

_When multiple sets of fastq files are in a 10x data bundle, the analysis service currently uses the lane_index to determine which R1, R2 and I1 files belong together. However, as @mckinsel brought up in a discussion about analyzing files from the same library prep together, if two sets of files have the same lane index because they were run on different flow cells we would not be able to determine which R1, R2 and I1 files belong together. I'm not sure if this is the case in any of the 10x datasets we currently have, but that could help determine how soon this is needed._

**Comment by ESapenaVentura** [comment]
I think this is outdated @clairerye

**Comment by hannes-ucsc** [comment]
This was addressed in the DCP/2.

Last paragraph of https://github.com/HumanCellAtlas/dcp2/blob/main/docs/dcp2_system_design.rst#data-file-paths
